### PR TITLE
50 favorite folders

### PR DESCRIPTION
**Pull request by MohammedAlghazali**
this pull request for user favorite folders (In the left side of the below picture)

this PR divided into parts
### add & modify & delete for spicific favorite folder
1- route **POST** ```/favorite/folder``` that add new favorite folder for specific user
- [x] function ```addFavoriteFolder```
- [x] query ```addFavoriteFolderQuery```
- [x] tests for route and query

2- route **PUT** ```/favorite/folder``` that edit the title of exist favorite folder for specific user
- [x] function ```editFavoriteFolder```
- [x] query ```editFavoriteFolderQuery```
- [x] tests for route and query

3- route **DELETE** ```/favorite/folder/:folderId``` that delete specific favorite folder for specific user
- [x] function ```deleteFavoriteFolder```
- [x] query ```deleteFavoriteFolderQuery```
- [x] tests for route and query

### add & remove courses from favorite folder
1- route **POST** ```/favorite/add-to-folder``` that add new course to specific favorite folder
- [x] function ```updateCourseToFolder```
- [x] query ```addCourseToFolderQuery```
- [x] tests for route and query

2- route **DELETE** ```/favorite/folder/:folderId/:courseId``` that reomve a course from specific favorite folder
- [x] function ```removeCourseFromFolder```
- [x] query ```deleteCourseFromFolderQuery```
- [x] tests for route and query

### CODE REVIEW

**Comment by AlaaTaima** [comment]
> this pull request for user favorite folders (In the left side of the below picture)
> 
> this PR divided into parts
> 
> ### add & modify & delete for spicific favorite folder
> 1- route **POST** `/favorite/folder` that add new favorite folder for specific user
> 
> * [x]  function `addFavoriteFolder`
> * [x]  query `addFavoriteFolderQuery`
> * [x]  tests for route and query
> 
> 2- route **PUT** `/favorite/folder` that edit the title of exist favorite folder for specific user
> 
> * [x]  function `editFavoriteFolder`
> * [x]  query `editFavoriteFolderQuery`
> * [x]  tests for route and query
> 
> 3- route **DELETE** `/favorite/folder/:folderId` that delete specific favorite folder for specific user
> 
> * [x]  function `deleteFavoriteFolder`
> * [x]  query `deleteFavoriteFolderQuery`
> * [x]  tests for route and query
> 
> ### add & remove courses from favorite folder
> 1- route **POST** `/favorite/add-to-folder` that add new course to specific favorite folder
> 
> * [x]  function `updateCourseToFolder`
> * [x]  query `addCourseToFolderQuery`
> * [x]  tests for route and query
> 
> 2- route **DELETE** `/favorite/folder/:folderId/:courseId` that reomve a course from specific favorite folder
> 
> * [x]  function `removeCourseFromFolder`
> * [x]  query `deleteCourseFromFolderQuery`
> * [x]  tests for route and query

**Good and clear description :+1:**

File: server/__test__/queries/editFavoriteFolderQuery.test.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+const connection = require('../../src/database/config/connection');
+const dbBuild = require('../../src/database/config/build');
+const { editFavoriteFolderQuery } = require('../../src/database/queries');
+
+describe('addCourseToFolderQuery', () => {
+  beforeAll(() => dbBuild());
+  afterAll(() => connection.end());
+
+  it('add Course To favorite Folder for specific user', async () => {
```
Comment (AlaaTaima) [inline server/__test__/queries/editFavoriteFolderQuery.test.js:9]: **fix the description here,**
```suggestion
  it('edit favorite folder title', async () => {
```

File: server/__test__/routes/addFavoriteFolder.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  afterAll(() => connection.end());
+
+  it('user add new favorite folder successfully', async () => {
+    expect.assertions(1);
+    const { body } = await request(app)
+      .post('/api/v1/favorite/folder')
+      .send({ title: 'first folder for user 1' })
+      .set('Cookie', userOneToken)
+      .expect(200);
+    expect(body).toStrictEqual({ message: 'folder add successfully' });
```
Comment (AlaaTaima) [inline server/__test__/routes/addFavoriteFolder.test.js:22]: **Fix the message content :** 
```suggestion
    expect(body).toStrictEqual({ message: 'new folder is added successfully' });
```

File: server/__test__/routes/addFavoriteFolder.test.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  });
+
+  it('un valid title in request body', async () => {
```
Comment (AlaaTaima) [inline server/__test__/routes/addFavoriteFolder.test.js:25]: **fix the description here,**
```suggestion
  it('in-valid title in request body', async () => {
```

File: server/__test__/routes/deleteFavoriteFolder.test.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  beforeAll(() => dbBuild());
+  afterAll(() => connection.end());
+
+  it('users should be able to delete favorite folder', async () => {
+    expect.assertions(1);
+    const { body } = await request(app)
+      .delete('/api/v1/favorite/folder/2')
+      .set('Cookie', userOneToken)
+      .expect(200);
+    expect(body).toStrictEqual({ message: 'folder deleted successfully' });
```
Comment (AlaaTaima) [inline server/__test__/routes/deleteFavoriteFolder.test.js:21]: **fix message here,** 
```suggestion
    expect(body).toStrictEqual({ message: 'folder has deleted successfully' });
```

File: server/__test__/routes/deleteFavoriteFolder.test.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  });
+
+  it('if user do not have this folder', async () => {
```
Comment (AlaaTaima) [inline server/__test__/routes/deleteFavoriteFolder.test.js:24]: **another typo mistake :**  
```suggestion
  it('if user does not has a specific  folder', async () => {
```

File: server/__test__/routes/deleteFavoriteFolder.test.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    const { body } = await request(app)
+      .delete('/api/v1/favorite/folder/test')
+      .set('Cookie', userOneToken)
+      .expect(400);
+    expect(body).toStrictEqual({
+      message: 'invalid inputs..!',
+    });
+  });
+
+  it('if user do not have token', async () => {
```
Comment (AlaaTaima) [inline server/__test__/routes/deleteFavoriteFolder.test.js:46]: fix typo here, 
```suggestion
  it('if user does not has a token', async () => {
```

File: server/src/controllers/favoriteFolder/deleteFavoriteFolder/index.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  const { folderId } = req.params;
+  const { id } = req.user;
+
+  try {
+    await folderIdSchema.validate({ folderId });
+    const integerFolderId = Number(folderId);
+    const { rows } = await checkUserFolder(id);
+    const isUserHaveFolder = rows.some(
+      (folder) => folder.id === integerFolderId
+    );
```
Comment (AlaaTaima) [inline server/src/controllers/favoriteFolder/deleteFavoriteFolder/index.js:17]: **Good choice of using `some` here :+1:**

**Comment by MohammedAlghazali** [comment]
@AlaaTaima Thank you very much

**Pull request merged**